RFC: DLpack support for interoperability with other GPU frameworks

EvenOldridge opened the pull request:
Comment period is open until 2019-12-13.

# DLpack support for interoperability with other GPU frameworks

| Status        | (Proposed)       |
:-------------- |:---------------------------------------------------- |
| **RFC #**     | [180](https://github.com/tensorflow/community/pull/180) |
| **Author(s)** | eoldridge@nvidia.com, wmjlyjemaine@gmail.com, zhoujinjing09@gmail.com |
| **Sponsor**   | apassos@google.com, sanjoy@google.com                 |
| **Updated**   | 2019-11-26                                           |

## Objective

This document proposes the adoption of [dlpack](https://github.com/dmlc/dlpack) as way of passing tensor data to other frameworks without leaving the GPU and without a copy per [24453](https://github.com/tensorflow/tensorflow/issues/24453).  dlpack is a community effort to define a common tensor data structure that can be shared by different frameworks. dlpack is currently supported by cuPy, cuDF, DGM, TGL, PyTorch, and MxNet. 

The interoperability of dlpack would allow for fast on-GPU communication between TensorFlow and these frameworks opening up a wide range of use cases outlined below.  It would further enable \_\_cuda_array_interface\_\_ interoperability through cuPy/cuDF which support both methods providing a way to transfer data to Numba, PyArrow and other frameworks that have adopted that method, although [a similar request has been made to support that method of interoperability](https://github.com/tensorflow/tensorflow/issues/29039) and ideally both would be supported.

alextp commented:
@sanjoy

ewilderj commented:
cc/ @brijk7

Thanks @EvenOldridge  -- we'll need you to have signed the CLA in order to proceed. If you believe you have and that the bot is in error, let me know and I can bump this through.

One nit: do @alextp and @sanjoy want their personal emails in the RFC, vs their work ones?

To @alextp and @sanjoy: thanks for sponsoring this. You'll be responsible for booking the design review meeting as usual, with the change that it will include the RFC authors (plus any other suggested community members) as well as the TF team.

Once the CLA is sorted out, @brijk7 can push forward with announcing and setting a review period.

alextp commented:
Work email is best I think.

On Mon, Nov 25, 2019 at 12:31 PM Edd Wilder-James <notifications@github.com>
wrote:

> cc/ @brijk7 <https://github.com/brijk7>
>
> Thanks @EvenOldridge <https://github.com/EvenOldridge> -- we'll need you
> to have signed the CLA in order to proceed. If you believe you have and
> that the bot is in error, let me know and I can bump this through.
>
> One nit: do @alextp <https://github.com/alextp> and @sanjoy
> <https://github.com/sanjoy> want their personal emails in the RFC, vs
> their work ones?
>
> To @alextp <https://github.com/alextp> and @sanjoy
> <https://github.com/sanjoy>: thanks for sponsoring this. You'll be
> responsible for booking the design review meeting as usual, with the change
> that it will include the RFC authors (plus any other suggested community
> members) as well as the TF team.
>
> Once the CLA is sorted out, @brijk7 <https://github.com/brijk7> can push
> forward with announcing and setting a review period.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/community/pull/180?email_source=notifications&email_token=AAABHRN5WFV3RPDVIISSTDTQVQYZFA5CNFSM4JRNFGZKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFDWMSY#issuecomment-558327371>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHRMKN2QLT3HN4K6UR3DQVQYZFANCNFSM4JRNFGZA>
> .
>


-- 
 - Alex

EvenOldridge commented:
@googlebot I signed it!

jermainewang commented:
@EvenOldridge Thanks for the RFC. Shall @VoVAllen and I sign the CLA as well?

VoVAllen commented on rfcs/20191016-dlpack-support.md:
```diff
@@ -0,0 +1,93 @@
+How it works?
+The first design consideration is that we want to avoid any modification to the main Tensorflow library, so to get around the potential long delay of PR, code review, and release cycle of Tensorflow main package. Inspired by the solution from https://github.com/tobegit3hub/tftvm, we decide to implement the functionality as two custom tensor ops: to_dlpack and from_dlpack.
+
+Besides, we want this feature to be plugged into other projects quite easily. For example, any project that relies on this feature is able to run without compiling against Tensorflow's header files. Not only that an extra dependency usually means extra effort, but also that such maintenance is repetitive and should be handled by the feature developer (i.e., us) alone. To this end, we have an idea ...
```
Yes. We hope this could be part of tensorflow. Currently we are using lots of workaround to prototype, which is not ideal. However, we need input from tensorflow people to see whether there's a more elegant way to do this.

EvenOldridge replied:
I've revised the RFC to be more clear on this.

VoVAllen commented on rfcs/20191016-dlpack-support.md:
```diff
@@ -0,0 +1,93 @@
+Finally, to achieve the maximal efficiency, we want the conversion happens without memory copy.
+
+For to_dlpack, the returned DLPack tensor shares the same memory address of the input Tensorflow tensor and holds a reference to it. Upon the destruction of the DLPack tensor, it will dereference the Tensorflow tensor, so it can be collected by Tensorflow's memory management. (inspired by PyTorch's DLPack implementation).
+For from_dlpack, it first creates an allocator object (subclass Tensorflow's allocator interface) that holds the reference to the DLPack tensor. The AllocateRaw function directly returns the memory it holds without creating any new buffer. Upon destruction, the DeallocateRaw function just calls the deletor of the DLPack tensor. (inspired by Tensorflow's immutable_constant_op).
```
Basically `DeallocateRaw` won't delete the tensor, but dereferencing the buffer. The data races/data free issue is handled by the original framework which produces this tensor.

Detail: https://github.com/VoVAllen/tf-dlpack/blob/master/src/to_dlpack_kernel.cc#L66

sanjoy replied:
That can work, but we need to be clear on the contract that TF will unref the dlpack Tensor as soon as all uses have been enqueued, and won't want for the kernels to actually finish.  As long as this is part of dlpack's contract all is good.

byronyi replied:
Echo the concern here. Take an example as followed:

Image another framework that mirrors current TF design, say X.

1. ``TF_ToDLPackOp`` increments the TF ``TensorBuffer`` ref count, and the tensors produced by the upstream TF_Op is ready for use **in its own stream context**;
2. ``X_FromDLPackOp`` executes **in its own stream context**, and assume synchronization is properly carried out (we still miss this piece though);
3. A downstream Op in X consumes this DLPack tensor **in its own stream context**, and you called ``TF_DeallocateRaw`` which immediately decrements the ref count;
4. Before the downstream Op in X actually finishes execution, TF reuses the ``TensorBuffer`` and starting to write new data onto it, causing a data race.

How do you plan to solve this problem I mentioned above?

VoVAllen replied:
> Echo the concern here. Take an example as followed:
> 
> Image another framework that mirrors current TF design, say X.
> 
> 1. `TF_ToDLPackOp` increments the TF `TensorBuffer` ref count, and the tensors produced by the upstream TF_Op is ready for use **in stream context A**;
> 2. `X_FromDLPackOp` executes **in stream context B**;
> 3. A downstream Op in X consumes this DLPack tensor **in stream context B**, and you called `TF_DeallocateRaw` which immediately decrements the ref count;
> 4. TF reuses the `TensorBuffer` and starting to write new data onto it in **in stream context A**.
> 
> How do you plan to sync stream A in TF and stream B in X?

@byronyi 
Minjie's comment (https://github.com/tensorflow/community/pull/180#issuecomment-568426092) partially addressed your concern. You are right about the situation, as the operation across stream is not guaranteed by dlpack so far. One solution is to sync the stream producing tensor to ensure the memory is ready before downstream's execution, which you can see from [mxnet](https://github.com/apache/incubator-mxnet/blob/10903e81100bb02eee3ecc033f858b846311f763/python/mxnet/ndarray/ndarray.py#L4877).

VoVAllen commented:
@EvenOldridge Thanks for the RFC.

brijk7 commented:
@EvenOldridge : once you update emails of sponsors (and address any initial comments), I will announce the RFC to developers@tensorflow.org. 
thanks!
brijesh

VoVAllen commented:
@EvenOldridge My email is zhoujinjing09@gmail.com if needed. Thanks!

EvenOldridge commented:
Thanks brijesh, I think the most recent update addresses the concerns and clarifies things a little.  Let me know if there are further questions or concerns.  Also big thanks to @VoVAllen and @jermainewang for their work on the tfdlpack package and for coauthoring.

EvenOldridge commented on rfcs/20191016-dlpack-support.md:
```diff
@@ -0,0 +1,93 @@
+:-------------- |:---------------------------------------------------- |
+| **RFC #**     | [NNN](https://github.com/tensorflow/community/pull/NNN) (update when you have community PR #)|
+| **Author(s)** | eoldridge@nvidia.com, wmjlyjemaine@gmail.com, @VoVAllen |
+| **Sponsor**   | alexandre.tp@gmail.com, sanjoy@playingwithpointers.com                 |
```
changed in the latest revision

EvenOldridge commented:
@ewilderj @brijk7 Just checking in to make sure there's nothing on my end holding this up.  I'm guessing it's due to the US holidays, but I want to make sure you don't need anything more from me at this juncture.

brijk7 commented:
@EvenOldridge : I have announced the RFC to developers@tensorflow.org for comments. Comment period ends on 12/13/19 (2 weeks). After incorporating feedback, you should be good to schedule a design review and move forward if there are no major points of contention.

sanjoy commented on rfcs/20191016-dlpack-support.md:
```diff
@@ -0,0 +1,114 @@
+Finally, to achieve the maximal efficiency, we want the conversion happens without memory copy.
+
+For to_dlpack, the returned DLPack tensor shares the same memory address of the input Tensorflow tensor and holds a reference to it. Upon the destruction of the DLPack tensor, it will dereference the Tensorflow tensor, so it can be collected by Tensorflow's memory management. (inspired by PyTorch's DLPack implementation).
+For from_dlpack, it first creates an allocator object (subclass Tensorflow's allocator interface) that holds the reference to the DLPack tensor. The AllocateRaw function directly returns the memory it holds without creating any new buffer. Upon destruction, the DeallocateRaw function just calls the deletor of the DLPack tensor. (inspired by Tensorflow's immutable_constant_op).
```
If you're in TensorFlow, you don't need to go through this indirection.  You could subclass `tensorflow::TensorBuffer` to implement the behavior you want.

VoVAllen replied:
The main reason uses allocator is due to the alignment issue. However I found the alignment requirement is always 64 bytes later. Will this change in the future? 
Also I agree`TensorBuffer` is simpler and do you think it's fine to always use 64 bytes alignment requirement in dlpack operation?

sanjoy replied:
I don't see how alignment is related to using `TensorBuffer` or not.  IIUC in from_dlpack you roughly do this:

```
if (IsAligned(64, dlm_tensor->dl_tensor.data)) {
  // forward dlpack data without memcpy
} else {
  // allocate output buffer and memcpy
}
```

This seems correct to me and we'll continue doing this when we use a `TensorBuffer` subclass.

VoVAllen replied:
You are right. I realized alignment is always 64 bytes after I finished my first version which used allocator.

sanjoy commented on rfcs/20191016-dlpack-support.md:
```diff
@@ -0,0 +1,93 @@
+How it works?
+The first design consideration is that we want to avoid any modification to the main Tensorflow library, so to get around the potential long delay of PR, code review, and release cycle of Tensorflow main package. Inspired by the solution from https://github.com/tobegit3hub/tftvm, we decide to implement the functionality as two custom tensor ops: to_dlpack and from_dlpack.
+
+Besides, we want this feature to be plugged into other projects quite easily. For example, any project that relies on this feature is able to run without compiling against Tensorflow's header files. Not only that an extra dependency usually means extra effort, but also that such maintenance is repetitive and should be handled by the feature developer (i.e., us) alone. To this end, we have an idea ...
```
If this lives within TF, I'd like to see a assessment of whether we can use [DT_VARIANT](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/variant.h) to hold instances of `DLManagedTensor`.

VoVAllen replied:
At python side, the dlpack is expected to be represented as a PyCapsule holding the address of DLManagedTensor. Therefore if I understand correctly about DT_VARIANT, it may not help in this case.

jermainewang replied:
@sanjoy Could you provide an example usage of DT_VARIANT? We'd like to investigate more about the proposal.

sanjoy replied:
Did you look at the [Variant class](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/variant.h) I linked above?  You should be able to look at its uses to find some examples.

sanjoy commented:
Sorry for the delay.  I'm happy with the overall direction.  I had two implementation concerns, both mentioned inline:

1. `FromDLPackOP` does not need to use TF allocator once it moves to live in TF.  It can instead subclass `TensorBuffer`.
2. I'd like to understand if we can represent `DLManagedTensor` as a DT_VARIANT instead of the [unsafe cast](https://github.com/VoVAllen/tf-dlpack/blob/4b305deab022b0e22c5335b0ea367752db16fad7/src/from_dlpack_kernel.cc#L88).

> @EvenOldridge : I have announced the RFC to [developers@tensorflow.org](mailto:developers@tensorflow.org) for comments. Comment period ends on 12/13/19 (2 weeks). After incorporating feedback, the sponsor should be good to schedule a design review and move forward if there are no major points of contention.

Does the 2020-01-07 Tuesday, 10:00am PT slot sound ok for the design review?

nouiz commented:
Could the subclass of tensorbuffer do a sync on the stream if DLPack doesn't support that?

Télécharger Outlook pour Android<https://aka.ms/ghei36>

________________________________
From: Sanjoy Das <notifications@github.com>
Sent: Tuesday, December 17, 2019 7:13:12 PM
To: tensorflow/community <community@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [tensorflow/community] RFC: DLpack support for interoperability with other GPU frameworks (#180)


@sanjoy commented on this pull request.

________________________________

In rfcs/20191016-dlpack-support.md<https://github.com/tensorflow/community/pull/180#discussion_r359095078>:

> +dlpack = thdlpack.to_dlpack(t2)  # th tensor -> dlpack
+t3 = tfdlpack.from_dlpack(dlpack)  # dlpack -> tf tensor
+print(t3)
+```
+You will find that t1, t2 and t3 all have the same values, shape, and device contexts.
+Package dependency: tensorflow>=2.0
+
+How it works?
+The first design consideration is that we want to avoid any modification to the main Tensorflow library, so to get around the potential long delay of PR, code review, and release cycle of Tensorflow main package. Inspired by the solution from https://github.com/tobegit3hub/tftvm, we decide to implement the functionality as two custom tensor ops: to_dlpack and from_dlpack.
+
+Besides, we want this feature to be plugged into other projects quite easily. For example, any project that relies on this feature is able to run without compiling against Tensorflow's header files. Not only that an extra dependency usually means extra effort, but also that such maintenance is repetitive and should be handled by the feature developer (i.e., us) alone. To this end, we have an idea of releasing it as a python package. However, the question is how to invoke the two custom tensor ops in python? The challenge is that Tensorflow's custom op interface has a limited support of argument and return types, while to_dlpack and from_dlpack should have an argument/return type of DLPack object. We work around this by encoding the address of an DLPack object as an integer, so it can be accepted/returned by the custom op interface. Then, we decode it in python or C depending on whether we return it (to_dlpack) or consume it (from_dlpack).
+
+Finally, to achieve the maximal efficiency, we want the conversion happens without memory copy.
+
+For to_dlpack, the returned DLPack tensor shares the same memory address of the input Tensorflow tensor and holds a reference to it. Upon the destruction of the DLPack tensor, it will dereference the Tensorflow tensor, so it can be collected by Tensorflow's memory management. (inspired by PyTorch's DLPack implementation).
+For from_dlpack, it first creates an allocator object (subclass Tensorflow's allocator interface) that holds the reference to the DLPack tensor. The AllocateRaw function directly returns the memory it holds without creating any new buffer. Upon destruction, the DeallocateRaw function just calls the deletor of the DLPack tensor. (inspired by Tensorflow's immutable_constant_op).


That can work, but we need to be clear on the contract that TF will unref the dlpack Tensor as soon as all uses have been enqueued, and won't want for the kernels to actually finish. As long as this is part of dlpack's contract all is good.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/tensorflow/community/pull/180?email_source=notifications&email_token=AABMF6YSHAAX7T7Q2G3ARQLQZFTJRA5CNFSM4JRNFGZKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCPRUQ5I#discussion_r359095078>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AABMF63WPS7HZCW4T6SHLRLQZFTJRANCNFSM4JRNFGZA>.

-----------------------------------------------------------------------------------
This email message is for the sole use of the intended recipient(s) and may contain
confidential information.  Any unauthorized review, use, disclosure or distribution
is prohibited.  If you are not the intended recipient, please contact the sender by
reply email and destroy all copies of the original message.
-----------------------------------------------------------------------------------

sanjoy commented:
> Could the subclass of tensorbuffer do a sync on the stream if DLPack doesn't support that? 

This is what I had in mind [here](https://github.com/tensorflow/community/pull/180#discussion_r350497617).

We can do either (sync on the destructor or not), but let's be clear on the contract with dlpack first.

VoVAllen commented:
Current implementation using OpKernel ensures that memory is prepared when entering this kernel. I'm wondering what change is needed for sync when we move to implement this inside TF. Thanks.

EvenOldridge commented:
> Does the 2020-01-07 Tuesday, 10:00am PT slot sound ok for the design review?

I can make that time work.  @VoVAllen @jermainewang are you available?

VoVAllen commented:
@EvenOldridge @sanjoy How long would it usually be? Because @jermainewang and I are in Shanghai. 10am PT is 2am at Shanghai, I think I can make it if it won't last too long.

jermainewang commented:
> > Does the 2020-01-07 Tuesday, 10:00am PT slot sound ok for the design review?
> 
> I can make that time work. @VoVAllen @jermainewang are you available?

Unfortunately, that is not good for me (2am Beijing time). How about afternoon? Anytime after 4 pm PST is good for me.

sanjoy commented:
> Current implementation using OpKernel ensures that memory is prepared when entering this kernel. I'm wondering what change is needed for sync when we move to implement this inside TF. Thanks.

Which kernel specifically are you talking about?

When you enter `ToDLPackOP` (say) all you know is that the work to compute the input tensors has been *enqueued* onto the compute stream but they might not have finished yet.  You can enqueue more kernels that use the inputs onto the compute stream and that will work fine because these kernels will be properly ordered, but you can't enqueue kernels that use the inputs onto independent streams without some form of syncing.

byronyi commented:
@sanjoy that’s also my concern here. Borrowing GPU memory without careful consideration of syncing compute streams most likely end up with either inefficient or incorrect interoperability. Unfortunately I did not see much discussion on this aspect in the RFC or the DLPack contract itself.

sanjoy commented:
To summarize, I'd like the authors to address the following points _before_ the design review:

1. What is the contract with DLPack around streams?
2. Can `FromDLPackOP` subclass `TensorBuffer` instead of creating a new allocator?
3. Can we represent `DLManagedTensor` as a `DT_VARIANT` to avoid the [unsafe cast](https://github.com/VoVAllen/tf-dlpack/blob/4b305deab022b0e22c5335b0ea367752db16fad7/src/from_dlpack_kernel.cc#L88)?

jermainewang commented:
Let me first clarify some details about DLPack and then answers your questions. DLPack itself is just a protocol of tensor data structure. To make it compatible with different frameworks that may have their own memory management mechanism, DLPack uses [`DLManagedTensor`](https://github.com/dmlc/dlpack/blob/master/include/dlpack/dlpack.h#L156) which contains a function pointer and its closure for memory deallocation. Both the function and closure are opaque pointers so that anything can be registered. Unfortunately, there is no formal contract at the moment. @tqchen please comment here. From my understanding, the contract is:

* In `to_dlpack`, the source framework creates a tensor using its own memory allocator and register the deallocator using `deleter` and `managed_ctx`.
* In `from_dlpack`, the target framework converts the DLPack tensor to its own tensor format and **makes sure that its memory is recycled by the registered `deleter`**.
* When the target framework acquires the memory block from a DLPack tensor, it assumes that **the memory is ready for read or write**.
* Once a DLPack tensor is converted to a tensor by the target framework (thus the memory has been *owned* by the target framework), the DLPack tensor is marked as **consumed** to avoid double ownership.

There is also a common protocol in the python side which is not explicitly documented. A DLPack pointer shall be wrapped in a python capsule containing the following members:
* A raw pointer to the DLPack object.
* A string flag indicating whether the object has been consumed or not. `'dltensor'` means unused capsule, while `'used_dltensor'` means consumed capsule.
* A deleter that calls the registered deleter of the underlying DLPack object.

This ensures that when the python capsule object goes out of scope, the DLPack object is correctly de-referenced.

> To summarize, I'd like the authors to address the following points _before_ the design review:
> 
> 1. What is the contract with DLPack around streams?

`to_dlpack` should make sure that the memory is ready for use. Therefore, a synchronization is probably required. However, all current adapters such as MXNet, PyTorch and DGL do not follow this. Another potential solution is to include the stream context in DLPack. I'd like to hear @tqchen 's thought on this.

> 2. Can `FromDLPackOP` subclass `TensorBuffer` instead of creating a new allocator?

We have thought about this approach. At that time, we found that a DLPack tensor may not contain aligned address, in which case we need to [reallocate the memory](https://github.com/VoVAllen/tf-dlpack/blob/a1fdb53096158c2ec9189bb1ff46c92c6f571bbe/src/from_dlpack_kernel.cc#L101-L121). I am not a hundred percent sure whether it is easy to implement that using `TensorBuffer`, but if it is the case, `TensorBuffer` is a much cleaner solution.

> 3. Can we represent `DLManagedTensor` as a `DT_VARIANT` to avoid the [unsafe cast](https://github.com/VoVAllen/tf-dlpack/blob/4b305deab022b0e22c5335b0ea367752db16fad7/src/from_dlpack_kernel.cc#L88)?

I think the answer is yes. We can represent a DLPack object using a DT_VARIANT tensor with variant type equal to void*. My question is how can I create such a tensor in python? I cannot find any documentation about that.

sanjoy commented:
> Let me first clarify some details about DLPack and then answers your questions. DLPack itself is just a protocol of tensor data structure. To make it compatible with different frameworks that may have their own memory management mechanism, DLPack uses [`DLManagedTensor`](https://github.com/dmlc/dlpack/blob/master/include/dlpack/dlpack.h#L156) which contains a function pointer and its closure for memory deallocation. Both the function and closure are opaque pointers so that anything can be registered. Unfortunately, there is no formal contract at the moment. @tqchen please comment here. From my understanding, the contract is:
> 
> * In `to_dlpack`, the source framework creates a tensor using its own memory allocator and register the deallocator using `deleter` and `managed_ctx`.
> * In `from_dlpack`, the target framework converts the DLPack tensor to its own tensor format and **makes sure that its memory is recycled by the registered `deleter`**.
> * When the target framework acquires the memory block from a DLPack tensor, it assumes that **the memory is ready for read or write**.
> * Once a DLPack tensor is converted to a tensor by the target framework (thus the memory has been _owned_ by the target framework), the DLPack tensor is marked as **consumed** to avoid double ownership.

Acknowledged.

I think and you and I are on the same page here (as is obvious from the rest of your reply), but for folks following along casually I'd like to emphasize that memory allocation and stream semantics are very closely tied and they need to be considered together.

> There is also a common protocol in the python side which is not explicitly documented. A DLPack pointer shall be wrapped in a python capsule containing the following members:
> 
> * A raw pointer to the DLPack object.
> * A string flag indicating whether the object has been consumed or not. `'dltensor'` means unused capsule, while `'used_dltensor'` means consumed capsule.

Out of curiosity why is this a string (and not a boolean)?

> * A deleter that calls the registered deleter of the underlying DLPack object.
> 
> This ensures that when the python capsule object goes out of scope, the DLPack object is correctly de-referenced.

I understand what you mean here, but (here and in the RFC) I'd prefer avoiding "de-reference" since in C/C++ [it means](https://en.wikipedia.org/wiki/Dereference_operator) reading from or writing to a pointer.

As a concrete suggestion, how about "decref-ed"?  It is clumsy but also less ambiguous.

> > To summarize, I'd like the authors to address the following points _before_ the design review:
> > 
> > 1. What is the contract with DLPack around streams?
> 
> `to_dlpack` should make sure that the memory is ready for use. Therefore, a synchronization is probably required. However, some current adapters such as PyTorch and DGL do not follow this.

Is this documented somewhere?  I'd prefer avoiding defensive programming here.  If the sync is truly not needed by the spec then we shouldn't do it.

> MXNet, on the other hand, performs the synchronization correctly ([here](https://github.com/apache/incubator-mxnet/blob/10903e81100bb02eee3ecc033f858b846311f763/python/mxnet/ndarray/ndarray.py#L4877) and [here](https://github.com/apache/incubator-mxnet/blob/10903e81100bb02eee3ecc033f858b846311f763/python/mxnet/ndarray/ndarray.py#L4909)). Another potential solution is to include the stream context in DLPack. I'd like to hear @tqchen 's thought on this.
> 
> > 1. Can `FromDLPackOP` subclass `TensorBuffer` instead of creating a new allocator?
> 
> We have thought about this approach. At that time, we found that a DLPack tensor may not contain aligned address, in which case we need to [reallocate the memory](https://github.com/VoVAllen/tf-dlpack/blob/a1fdb53096158c2ec9189bb1ff46c92c6f571bbe/src/from_dlpack_kernel.cc#L101-L121). I am not a hundred percent sure whether it is easy to implement that using `TensorBuffer`, but if it is the case, `TensorBuffer` is a much cleaner solution.

Maybe I'm missing something, but this is what I had in mind:

```
class DlPackTensorBuffer : public TensorBuffer {
  // ... Remembers the DLPack pointer and the deletion function
};


FromDlPackOp::Compute() {
  if (address is aligned) {
    ctx->set_output(Tensor(DlPackTensorBuffer(address, deleter)));
  } else {
    // allocate and memcpy as you do today.  Don't use DlPackTensorBuffer.
  }
}
```

> > 1. Can we represent `DLManagedTensor` as a `DT_VARIANT` to avoid the [unsafe cast](https://github.com/VoVAllen/tf-dlpack/blob/4b305deab022b0e22c5335b0ea367752db16fad7/src/from_dlpack_kernel.cc#L88)?
> 
> I think the answer is yes. We can represent a DLPack object using a DT_VARIANT tensor with variant type equal to void*. My question is how can I create such a tensor in python? I cannot find any documentation about that.

IMO using a `DT_VARIANT` tensor with variant type equal to `void*` does not add much value.  I was actually thinking of using `DT_VARIANT` tensors with variant type equal to `DLManagedTensor`.  However, I don't see an obvious way to do that so probably there's nothing to do here.

tqchen commented:
Re the need of sync. DLPack is a memory exchange protocol, so sync is not strictly necessary, if the consumer and the producer uses the same stream. MXNet's implementation did the sync mainly because its scheduler asynchrously assigned a potentially different stream. So in order for subsequent consumer to consume the content correctly, it does an internal sync.

In the case of TF, I can think of two possible scenarios:
- When TF's internal scheduler automatically picks different variations of streams, and we want to expose the DLPack Tensor to other framework(say pytorch), then a sync is necessary. However, if TF uses the default stream, then sync may not be necessary.

- When we expose another DLPack compatible function(e.g. those compiled by TVM) as a callback to implement the TF operator, and the stream is set by the TF runtime, then sync should not be performed.

sanjoy commented:
> Re the need of sync. DLPack is a memory exchange protocol, so sync is not strictly necessary, if the consumer and the producer uses the same stream.

Agreed.  However, IIUC the consumer and producers are different frameworks so they will usually use different streams.

> MXNet's implementation did the sync mainly because its scheduler asynchrously assigned a potentially different stream. So in order for subsequent consumer to consume the content correctly, it does an internal sync.
> 
> In the case of TF, I can think of two possible scenarios:
> 
> * When TF's internal scheduler automatically picks different variations of streams, and we want to expose the DLPack Tensor to other framework(say pytorch), then a sync is necessary. However, if TF uses the default stream, then sync may not be necessary.

TF does not use default streams as far as I know.  See [here](https://github.com/tensorflow/tensorflow/blob/c23505c657e6a40a84c4e983ea529a1310c4a6e1/tensorflow/core/common_runtime/gpu/gpu_device.cc#L247).

And even if it did, default streams are one per thread right?  The TF executor is multi-threaded so even if it uses default streams there will be a mismatch between it and other frameworks.

> * When we expose another DLPack compatible function(e.g. those compiled by TVM) as a callback to implement the TF operator, and the stream is set by the TF runtime, then sync should not be performed.

Sorry, I don't understand two of the terms you used: "DLPack compatible function", "callback to implement the TF operator".

tqchen commented:
OK, to rephrase my last post.

If the user has a kernel function that takes DLManagedTensor as arguments, and can be configured to use the stream that TF uses. Then if we want to use it to define a TF operator(say TFDLPackOp), then we do not need to sync, and should not in order to get the best efficiency.

On the other hand, if we want to export an array in TF to say Pytorch, and we cannot control the stream pytorch and TF uses, then it makes sense to sync.

So to support both cases, perhaps we will need two kinds of APIs

sanjoy commented:
> If the user has a kernel function that takes DLManagedTensor as arguments, and can be configured to use the stream that TF uses(by getting that from TF's operator context).

If the user is building their kernel function specifically for TF then why would they not write a `tnesorflow::OpKernel` that directly works on `tensorflow::Tensor`s?

tqchen commented:
> If the user is building their kernel function specifically for TF then why would they not write a 
   tensorflow::OpKernel that directly works on tensorflow::Tensors?

Think it as a fast adapter, e.g. that kernel could be written in other form that can take DLPack tensors,  e.g. pytorch functions, code that get automatically generated by a compiler which can work with a C ABI compatible data structure(DLTensor*) but not necessarily c++ structure

sanjoy commented:
The next action item is to evaluate whether the DLPack interface can and should be implemented using the `TFE_NewTensorHandleFromDeviceMemory`, `TFE_TensorHandleDeviceMemorySize`, `TFE_TensorHandleDevicePointer` and a hypothetical TBD `TFE_SynchronizeDevice` APIs.  Once we have made a decision on that, we can merge this RFC and start reviewing PRs.

**Design review meeting notes**

Sanjoy: 3 questions, 
  (1) what about streams? Resolution is that frameworks need to sync their streams at the boundaries

Minjie: we can make an official document from DLPack about what should happen around streams when handing tensors around

 (2) subclassing TensorBuffer vs using new allocator. Minjie: yes, we plan on doing this.

 (3) around using DT_VARIANT to represent DLManagedTensor. Apparently there's no easy way.

Minjie: no easy to way to pass a pointer from python to TensorFlow.
Alex: we can use TFE_NewTensorHandleFromDeviceMemory.
Minjie: when will this make it into a final release?
Alex: it's not in 2.1, will be in 2.2.
Minjie: then we might need an interim solution until then.

Sanjoy: but either way this feature won't be available in core TF till the 2.2 release.  And in the interim we have the tf-dlpack out of tree implementation.

Agarwal: is alignment a concern? Minjie: the tf to dlpack layer needs to copy the memory to ensure alignment.

Even Oldridge: can we throw an error on unaligned things? Alex: yes, we already get errors.

Agarwal: beware of int32 tensors and the host memory problems. Followed by a short discussion of the problems.

Sanjoy: should we mark any new ops as experimental?
Alex: yes, any new python APIs should probably be experimental for at least one release.

Sanjoy: what about cuda_array_interface? It looks like the device memory tensor handle API would be sufficient to implement that. We also need a C API to synchronize GPU streams.

Minjie: if TF exposed its streams in the API then integration problems would be made simpler. Maybe TF should just use the default stream?
Sanjoy: how would we represent this in the API?
Minjie: maybe a string opaque stream ID?
Sanjoy: need to make this work across cuda / rocm so hopefully some generic representation is possible.

Wolff: how can we make sure users actually know about this?
Resolution: general brainstorm about using TF twitter, dgl twitter, rapids twitter, etc.

byronyi commented:
Recently I saw @hawkinsp from the XLA team pulled in DLPack into TF core for interoperability between JAX and other GPU libraries in https://github.com/tensorflow/tensorflow/commit/fc1f6fdf943faf60b3127a1e1b0b8719a4ddaa32. Not sure if it's related to this RFC per se.

hawkinsp commented:
@byronyi That's right; I added DLPack support to the XLA:Python bindings for use in JAX (https://github.com/google/jax). That's somewhat separate to this RFC, which pertains to TensorFlow. However, since XLA and TensorFlow live in the same repository, the PR you linked did some preparatory work that may also assist with an implementation in TensorFlow (e.g., it adds DLPack and its headers to the TF bazel workspace.)

Apropos of the discussion about synchronization, I chose to synchronize the device to the host when converting a JAX array to a DLManagedTensor. This seemed like the safest choice, given DLPack doesn't have any mechanisms to describe device-side synchronization (e.g., a CUDA Event that describes when a particular buffer becomes valid on device would be an obvious choice in the case of CUDA.)

dynamicwebpaige commented:
Thanks again for submitting the RFC, @EvenOldridge, @jermainewang, and @VoVAllen!

Is there anything that we need to do to move this along? DLPack support is an exciting feature, and I want to make sure that we keep momentum. :slightly_smiling_face:

VoVAllen commented:
@sanjoy I created the modification PR  https://github.com/EvenOldridge/community/pull/2 to @EvenOldridge's branch.
@EvenOldridge Could you help merge this? Thanks!

sanjoy commented on rfcs/20191016-dlpack-support.md:
```diff
@@ -0,0 +1,130 @@
+to show how this design addresses the problem?
+
+Which users are affected by the problem? Why is it a problem? What data supports
+this? What related work exists?
```
These questions can be deleted right?

EvenOldridge replied:
Removed

sanjoy commented on rfcs/20191016-dlpack-support.md:
```diff
@@ -0,0 +1,130 @@
+## User Benefit
+
+How will users (or other contributors) benefit from this work? What would be the
+headline in the release notes or blog post?
```
These questions should be deleted as well.

EvenOldridge replied:
Removed

sanjoy commented on rfcs/20191016-dlpack-support.md:
```diff
@@ -0,0 +1,130 @@
+AFAICT it should be easy to take cuda pointers in and out of TF and use them to build dlpack structures from tensors or vice versa. The tricky part is that TF does not use cudamalloc to allocate memory but its own allocator whose internal state is stored on the CPU and matches the head of TF's compute stream, so we need to sync TF's stream before the memory is usable from dlpack and similarly syn ...
+
+A working version of dlpack integration has been released as a package by coauthors @jermainewang and @VoVAllen here:
+https://github.com/VoVAllen/tf-dlpack/issues/3
```
I would prefer rephrasing text like this and making the proposal describe what was actually implemented.

Same for other "as mentioned by alextp" comments below.

jermainewang commented on rfcs/20191016-dlpack-support.md:
```diff
@@ -0,0 +1,121 @@
+
+## Objective
+
+This document proposes the adoption of dlpack (https://github.com/dmlc/dlpack) as way of passing tensor data to other frameworks without leaving the GPU and without a copy per [24453](https://github.com/tensorflow/tensorflow/issues/24453).  dlpack is a community effort to define a common tensor data structure that can be shared by different frameworks. dlpack is currently supported by cuPy, cuDF, ...
```
spell error? should be DGL.

VoVAllen replied:
Fixed in the latest pr

VoVAllen commented:
I made a new pr to @EvenOldridge's branch. Hope this looks better.

sanjoy commented on rfcs/20191016-dlpack-support.md:
```diff
@@ -0,0 +1,122 @@
+
+DLPack is a community effort to define a common tensor data structure that can be shared by different frameworks allowing data to be quickly shared often with zero or minimal copy. One of the main bottlenecks when trying to achieve GPU performance when operating across different frameworks is I/O and data formatting.  The transfer of data between GPU and CPU or between formats is costly to the po ...
+
+Existing applications that take advantage of dlpack include: (adding my own and those listed in , other contributions needed)
```
`(adding my own and those listed in , other contributions needed)` seems malformed

VoVAllen commented:
https://github.com/EvenOldridge/community/pull/4
Added a new pr to address comment above @EvenOldridge
